### PR TITLE
Make NetworkPartition disruption scheme configurable

### PR DESCRIPTION
**Pull request by ywelsch**
We currently have subclasses for the various kind of network disruption schemes combining on one hand failure mode (disconnect/unresponsive/delay) as well as the network links to cut (two partitions / bridge partitioning) into a single class. This PR separates the description of the links in the network that are to be disrupted from the failure that is to be applied to the links (disconnect/unresponsive/delay).

### CODE REVIEW

**Comment by rmuir** [comment]
I like how the code consolidates these classes into one place where they can be more easily contained. 

Can we improve the javadocs just a tad more to make it more obvious what is happening?
* add class javadocs to the different subclasses. just a one-liner what it does, to make it obvious.
* add method javadocs to abstract methods (e.g. `applyDisruption()`) as those represent a contract.

**Comment by ywelsch** [comment]
Thanks for the feedback @rmuir. I've pushed c302837 adding more javadocs.

File: core/src/test/java/org/elasticsearch/cluster/MinimumMasterNodesIT.java [hunk 1]
```diff
@@ -372,12 +375,14 @@ public void testCanNotPublishWithoutMinMastNodes() throws Exception {
...
-        NetworkDelaysPartition partition = new NetworkDelaysPartition(Collections.singleton(master), otherNodes, 60000, random());
+        NetworkDisruption partition = new NetworkDisruption(
+            new TwoPartitions(Collections.singleton(master), otherNodes),
+            new NetworkDelay(TimeValue.timeValueMinutes(1)));
         internalCluster().setDisruptionScheme(partition);
         partition.startDisrupting();
 
         final CountDownLatch latch = new CountDownLatch(1);
-        final AtomicReference<Throwable> failure = new AtomicReference<>();
+        final AtomicReference<Exception> failure = new AtomicReference<>();
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/cluster/MinimumMasterNodesIT.java]: 🎉

File: test/framework/src/main/java/org/elasticsearch/test/disruption/NetworkDisruption.java [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+    }
+
+    @Override
+    public synchronized void applyToNode(String node, InternalTestCluster cluster) {
+
+    }
+
+    @Override
+    public synchronized void removeFromNode(String node1, InternalTestCluster cluster) {
+        if (disruptedLinks.nodes().contains(node1)) {
```
Comment (jasontedor) [inline test/framework/src/main/java/org/elasticsearch/test/disruption/NetworkDisruption.java]: Can you do something about the duplicated deeply nested code here and in `startDisrupting`?

**Comment by jasontedor** [comment]
I like it @ywelsch, it's a great change. I left one comment.

File: test/framework/src/main/java/org/elasticsearch/test/disruption/NetworkDisruption.java (hunk 2 above)
Comment (ywelsch) [inline test/framework/src/main/java/org/elasticsearch/test/disruption/NetworkDisruption.java]: I've pushed a72b22e to eliminate the code duplication.

**Comment by jasontedor** [comment]
LGTM.

**Pull request merged**